app/models/internal/scheduling/JobExecution.java

```diff
@@ -0,0 +1,435 @@
+     */
+    public interface Visitor<T, U> extends Function<JobExecution<T>, U> {
+        /**
+         * Convenience wrapper around {@code state.accept(visitor)}.
+         *
+         * @param state The state to visit.
+         * @return The result produced by this visitor.
+         */
+        @Override
+        default U apply(final JobExecution<T> state) {
+            return state.accept(this);
+        }
+
+        /**
+         * Visit a started state.
+         *
+         * @param state The state to visit.
+         * @return The result produced by this visitor.
+         */
+        U visit(JobExecution.Started<T> state);
+
+        /**
+         * Visit a success state.
+         *
+         * @param state The state to visit.
+         * @return The result produced by this visitor.
+         */
+        U visit(JobExecution.Success<T> state);
+
+        /**
+         * Visit a failure state.
+         *
+         * @param state The state to visit.
+         * @return The result produced by this visitor.
+         */
+        U visit(JobExecution.Failure<T> state);
+    }
+
+    /**
+     * Base builder class for all subclasses of {@code JobExecution}.
+     *
+     * @param <B> The concrete type of this builder.
+     * @param <T> The result produced by the output JobExecution.
+     * @param <U> The output of this builder.
+     */
+    protected abstract static class Builder<B extends Builder<B, T, U>, T, U extends JobExecution<T>> extends OvalBuilder<U> {
+        @NotNull
+        private UUID _jobId;
+        @NotNull
+        private Instant _scheduled;
+
+        /**
+         * Forwarding Constructor.
+         *
+         * @param targetConstructor The constructor for a {@code U}.
+         */
+        protected Builder(final Function<B, U> targetConstructor) {
+            super(targetConstructor);
+        }
+
+        /**
+         * Set the jobId for this execution.
+         *
+         * @param jobId The job Id.
+         * @return This instance of builder.
+         */
+        public B setJobId(@Nullable final UUID jobId) {
+            _jobId = jobId;
+            return self();
+        }
+
+        /**
+         * Set the scheduled time this execution.
+         *
+         * @param scheduled The instant this execution is scheduled for.
+         * @return This instance of builder.
+         */
+        public B setScheduled(@Nullable final Instant scheduled) {
+            _scheduled = scheduled;
+            return self();
+        }
+
+        /**
+         * Wrapper for propagating the concrete type of a derived builder instance.
+         *
+         * @return This instance of builder.
+         */
+        protected abstract B self();
+    }
+
+    /**
+     * An execution that has been started but not yet completed.
+     *
+     * @param <T> The result type of this execution.
+     */
+    public static final class Started<T> extends JobExecution<T> {
```

Comment:
Does this need to be generic?

Reply:
Yep - the type parameter is essentially a phantom type here. It's needed to be able to uniformly handle each variant of `JobExecution<T>` for a particular `T`. 

Otherwise you could assign a `JobExecution.Started<Integer>` to a `JobExecution<String>` which doesn't seem correct regardless of if the object actually contains a `T` or not.

A more pragmatic reason is that even if I wanted to leave it out the compiler would complain that the type variable `T` is not defined anywhere to be specified in the `extends` clause.